src/cargo/core/compiler/mod.rs

```diff
@@ -1865,12 +1865,30 @@ fn on_stderr_line_inner(
         return Ok(true);
     }
 
-    #[derive(serde::Deserialize)]
-    struct CompilerMessage {
-        level: String,
-    }
-    if let Ok(message) = serde_json::from_str::<CompilerMessage>(compiler_message.get()) {
-        count_diagnostic(&message.level, options);
+    if let Ok(msg) = serde_json::from_str::<CompilerMessage>(compiler_message.get()) {
+        if msg.message.starts_with("aborting due to")
+            || msg.message.ends_with("warning emitted")
+            || msg.message.ends_with("warnings emitted")
+        {
+            // Skip this line; we'll print our own summary at the end.
+            return Ok(true);
+        }
+        let rendered = msg.rendered;
+        if options.show_diagnostics {
+            let machine_applicable: bool = msg
+                .children
+                .iter()
+                .map(|child| {
+                    child
+                        .spans
+                        .iter()
+                        .filter_map(|span| span.suggestion_applicability)
+                        .any(|app| app == Applicability::MachineApplicable)
+                })
+                .any(|b| b);
+            count_diagnostic(&msg.level, options);
+            state.emit_diag(msg.level, rendered, machine_applicable)?;
```

Comment:
I am not sure if we need to emit diagnostics here. Could you give an example for the necessity of this?

(The example can be done as a test case, if applicable)

Reply:
I'm not sure whether this diagnostic is expected or not, but in issue #14472, running `cargo check` will emit a warning, but `cargo check --message-format json` won't emit a warning. Do you think this should be aligned?

Reply:
The message format is explicitly chosen as JSON, so I don't expect any message to be emitted in a human-readable format.